### PR TITLE
Block

### PR DESCRIPTION
**Pull request by zhfnjust**

### CODE REVIEW

File: contracts/merklePath.scrypt [hunk 1]
```diff
@@ -0,0 +1,42 @@
+
+struct Sibling {
```
Comment (xhliu) [inline contracts/merklePath.scrypt:2]: Sibling -> Node

Let's use node to be consistent w/ the community
https://github.com/bitcoin-sv-specs/merkle-proof-standard-example/blob/4f4b47149bc52c292328c3c9b114b8a919560d7a/verifyMerkleProofJSON.js#L63

File: contracts/merklePath.scrypt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    Sha256 hash;
+    // invalid {0} or left {1} or right {2} sibling
```
Comment (xhliu) [inline contracts/merklePath.scrypt:4]:
```
// 0: invalid
// 1: on the left
// 2: on the right
```

File: contracts/merklePath.scrypt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    int left;
+}
+
+type MerklePaths = Sibling[MerklePath.DEPTH];
+
+library MerklePath {
+    // N is length of a Sibling
+    static const int N = 33;
+    // can support a block with 4294967296 transations
+    static const int DEPTH = 32;
```
Comment (xhliu) [inline contracts/merklePath.scrypt:14]: Replace 0/1/2 w/ these constants.

```
static const int INVALID_NODE = 0;
static const int LEFT_NODE = 1;
static const int RIGHT_NODE = 2;
```

File: contracts/merklePath.scrypt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    //According to the given leaf node and merkle path, calculate the hash of the root node of the merkle tree
+    static function calMerkleRoot(Sha256 leaf, MerklePaths merklePath) : Sha256 {
+        Sha256 root = leaf;
+
+        loop (DEPTH) : i {
+            Sibling s = merklePath[i];
+            if(s.left > 0) { // s is valid
```
Comment (xhliu) [inline contracts/merklePath.scrypt:21]: Replace 0/1/2 w/ these constants. Everywhere.
`if (s.left != INVALID_NODE) `

File: contracts/blockchain.scrypt [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    bytes nonce;
+}
+
+
+// a library to trustlessly access the blockchain: including blockheaders and transactions
+library Blockchain {
+    // block height's position relative to the beginning of coinbase tx
+    // TODO: This assumes unlocking script can be pushed using OP_PUSH_1. See if it always holds?
+    static const int BLOCK_HEIGHT_POS = 42;
+    static const int DEPTH = 2;
```
Comment (xhliu) [inline contracts/blockchain.scrypt:21]: comment

File: contracts/blockchain.scrypt [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                    if (prevBlockHash != headers[i].prevBlockHash)
+                        res = false; 
+                }
+            }
+        }
+
+        return res;
+    }
+
+    // 01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff1d0352e40a2f7376706f6f6c2e636f6d2f306c92f79739e3174831970000ffffffff01836e4125000000001976a91468888423f412b6b4166d61fe9e66a1aeff30df1f88ac00000000
```
Comment (xhliu) [inline contracts/blockchain.scrypt:108]: remove

File: contracts/blockchain.scrypt [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        return tx[4:5] == b'01' // only 1 input
+            && tx[5:37] == b'0000000000000000000000000000000000000000000000000000000000000000' // null txid: all zeros
+            && tx[37:41] == b'FFFFFFFF';    // null vout: all Fs
+    }
+
+    // a coinbase tx is the 1st tx in a block; thus all siblings on the merkle path must be on the right
+    static function isCoinbase2(MerklePaths merklePath) : bool { 
+        return MerklePath.isCoinbase(merklePath);
+    }
+
```
Comment (xhliu) [inline contracts/blockchain.scrypt:120]: delete

File: contracts/blockchain.scrypt [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    // get height of the block identified by the header
+    static function blockHeight(BlockHeader bh, bytes coinbaseTx, MerklePaths merklePath) : int {
+        
+        // ensure coinbase it's in the block
+        require(txInBlock(hash256(coinbaseTx), bh, merklePath));
+
+        // ensure it's coinbase
+        require(isCoinbase(coinbaseTx));
+        // alternative
+        // require(isCoinbase2(merklePath));
```
Comment (xhliu) [inline contracts/blockchain.scrypt:130]: Just use 
MerklePath.isCoinbase

File: contracts/blockchain.scrypt [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+    // block header hash
+    static function blockHeaderHash(BlockHeader bh) : Sha256 {
+        return hash256(serialize(bh));
+    }
+
+    // block header hash, but converted to a number
+    static function blockHeaderHashAsInt(BlockHeader bh) : int {
+        // hash is reversed
+        // no need to change to unsigned here since it is already positive, due to leading zeros
```
Comment (xhliu) [inline contracts/blockchain.scrypt:163]: delete

File: contracts/blockchain.scrypt [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    static function serialize(BlockHeader bh) : bytes {
+        return bh.version + bh.prevBlockHash + bh.merkleRoot + bh.time + bh.bits + bh.nonce;
+    }
+
+    // block header hash
+    static function blockHeaderHash(BlockHeader bh) : Sha256 {
+        return hash256(serialize(bh));
+    }
+
+    // block header hash, but converted to a number
```
Comment (xhliu) [inline contracts/blockchain.scrypt:160]: to a positive number

File: contracts/blockchain.scrypt [hunk 11]
```diff
@@ -0,0 +1,166 @@
+import "util.scrypt";
+import "merklepath.scrypt";
+
+
+struct BlockHeader {
+    bytes version;
+    Sha256 prevBlockHash;
+    Sha256 merkleRoot;
+    bytes time;
```
Comment (xhliu) [inline contracts/blockchain.scrypt:9]: `int time;`

File: contracts/blockTimeBet.scrypt [hunk 12]
```diff
@@ -0,0 +1,35 @@
+import "blockchain.scrypt";
+
+// bet on block time: how long it takes to mine the block containing the bet transaction
+contract BlockTimeBet {
+    // only 2 is needed; 7 means the transaction has 6 confirmations
+    static const int N = 7;
+    // 10 minutes in seconds
+    static const int AVG_BLOCK_TIME = 600;
+
+    int blockchainTarget;
```
Comment (xhliu) [inline contracts/blockTimeBet.scrypt:10]: add a comment
`// maximal target for any block to be considered valid`

File: contracts/blockchainTest.scrypt [hunk 13]
```diff
@@ -0,0 +1,38 @@
+import "blockchain.scrypt";
+
+
+contract BlockchainTest {
+    
+    int blockchainTarget;
+    
+    public function testBlockHeight(BlockHeader bh, MerklePaths merklePath,  bytes coinbaseTx, int height, SigHashPreimage txPreimage) {
```
Comment (xhliu) [inline contracts/blockchainTest.scrypt:8]: height -> expectedHeight

File: contracts/blockchainTest.scrypt [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        require(Tx.checkPreimage(txPreimage));
+
+        // get id of previous tx
+        Sha256 prevTxid = Sha256(Util.outpoint(txPreimage)[:32]);
+
+        // validate block header
+        require(Blockchain.isBlockHeaderValid(bh, this.blockchainTarget));
+
+        // verify previous tx is in the block
+        require(Blockchain.txInBlock(prevTxid, bh, merklePath));
```
Comment (xhliu) [inline contracts/blockchainTest.scrypt:18]: These are not relevant to height. Put them in separate functions if u wanna test them.

File: contracts/blockchainTest.scrypt [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+        require(Blockchain.blockTxCount(bh, lastTxid, merklePathOfLastTx)  == count);
```
Comment (xhliu) [inline contracts/blockchainTest.scrypt:36]: Same as above

File: tests/js/blockPRNG.scrypttest.js [hunk 16]
```diff
@@ -0,0 +1,185 @@
+const { expect } = require('chai');
+const { bsv, buildContractClass, PubKey, toHex, Sha256, Bytes, getPreimage, signTx, buildTypeClasses, num2bin } = require('scryptlib');
+const { num2hex, sha256d, reverseEndian, compileContract, inputIndex, inputSatoshis, uint322bin, toLittleIndian, pdiff2Target } = require('../../helper');
+const BN = bsv.crypto.BN
+
+function newTx() {
```
Comment (xhliu) [inline tests/js/blockPRNG.scrypttest.js:6]: Why is this shared code not in helper?

File: tests/js/blockchainTest.scrypttest.js [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+        txId: 'c1d32f28baa27a376ba977f6a8de6ce0a87041157cef0274b20bfda2b0d8df96',
+        outputIndex: 0,
+        script: '',   // placeholder
+        satoshis: inputSatoshis
+    };
+    return new bsv.Transaction().from(utxo);
+}
+
+// COINBASE TX of block 575191
+let COINBASETX = '01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff1c03d7c6082f7376706f6f6c2e636f6d2f3edff034600055b8467f0040ffffffff01247e814a000000001976a914492558fb8ca71a3591316d095afc0f20ef7d42f788ac00000000';
```
Comment (xhliu) [inline tests/js/blockchainTest.scrypttest.js:17]: let -> const

Use const everywhere for constants

File: tests/js/blockchainTest.scrypttest.js [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+            "hash": "66e3dee0ace4efd7b4f5406ea63d5bfb5703ba35ddd4c091e8fb2a99a707fe3e",
+            "pos": "L"
+        }
+    ],
+    "hash": "5c4d44b9b8d2ec6e0835ac90f206cecb26bf51033f31d4c659975b7534853409",
+    "merkleRoot": "95a920b1002bed05379a0d2650bb13eb216138f28ee80172f4cf21048528dc60"
+}
+
+
+// it's the secound tx in the block
```
Comment (xhliu) [inline tests/js/blockchainTest.scrypttest.js:75]: secound -> second

File: tests/js/blockchainTest.scrypttest.js [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+let NotLastTxID = '7e0ba1980522125f1f40d19a249ab3ae036001b991776813d25aebe08e8b8a50';
+// merklePath of secound tx in the block
```
Comment (xhliu) [inline tests/js/blockchainTest.scrypttest.js:77]: secound

File: tests/js/blockchainTest.scrypttest.js [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+describe('Test sCrypt contract blockchainTest In Javascript', () => {
+    const BlockchainTest = buildContractClass(compileContract('blockchainTest.scrypt'))
+
+    const { BlockHeader, Sibling } = buildTypeClasses(BlockchainTest);
+
+    function buildMerklePath(merklePath) {
+        const array = new Array(32);
+        array.fill(new Sibling({
+            hash: new Sha256("0000000000000000000000000000000000000000000000000000000000000000"),
+            left: 0 //invalid sibling
```
Comment (xhliu) [inline tests/js/blockchainTest.scrypttest.js:124]: Also change 0/1/2 to INVALID/LEFT/RIGHT for readability in JS

File: helper.js [hunk 21]
```diff
@@ -219,6 +217,81 @@ async function fetchUtxos(address) {
...
+
+function toLittleIndian(hexstr) {
+  return reverseEndian(hexstr)
+}
+
+function toBigIndian(hexstr) {
+  return reverseEndian(hexstr)
+}
+
+function uint322bin(d) {
```
Comment (xhliu) [inline helper.js:236]: uint322bin -> uint32Tobin

File: helper.js [hunk 22]
```diff
@@ -219,6 +217,81 @@ async function fetchUtxos(address) {
...
+  if (typeof difficulty === 'number') {
+      difficulty = BigInt(Math.floor(difficulty))
+  }
+
+  return BigInt(toTarget("1d00ffff") / difficulty);
+}
+
+
+// concat Header to get raw header
+function concatHeader(header) {
```
Comment (xhliu) [inline helper.js:284]: concatHeader -> serializeHeader

File: tests/js/blockchainTest.scrypttest.js [hunk 23]
```diff
@@ -0,0 +1,266 @@
...
+            nonce: new Bytes(uint322bin(header.nonce))
+        })
+    }
+
+    before(() => {
+        //Normally, the difficulty of the current network should be used
+        blockchainTest = new BlockchainTest(pdiff2Target(header.difficulty))
+    })
+
+    it('blockchainTest should succeed when using right block header', () => {
```
Comment (xhliu) [inline tests/js/blockchainTest.scrypttest.js:156]: add cases where 

- header validation fails due to low difficulty
- wrong block height

File: tests/js/blockchainTest.scrypttest.js [hunk 24]
```diff
@@ -0,0 +1,266 @@
...
+        tx.addOutput(new bsv.Transaction.Output({
+            script: bsv.Script.buildPublicKeyHashOut(privateKey.toAddress()),
+            satoshis: outputAmount
+        }));
+
+
+        blockchainTest.txContext = { tx, inputIndex, inputSatoshis }
+
+
+        preimage = getPreimage(tx, blockchainTest.lockingScript, inputSatoshis)
```
Comment (xhliu) [inline tests/js/blockchainTest.scrypttest.js:187]: Looks like repeated code across `it()`s. Put in a wrapper function.

File: tests/js/blockchainTest.scrypttest.js [hunk 25]
```diff
@@ -0,0 +1,266 @@
...
+            buildMerklePath(merklePathOfLastTx),
+            new Sha256(toLittleIndian(Last_TX_ID)),
+            3,
+            preimage).verify()
+
+        expect(result.success, result.error).to.be.true
+    });
+
+
+    it('testBlockTxCount should FAIL when using last tx copy in the merkle tree', () => {
```
Comment (xhliu) [inline tests/js/blockchainTest.scrypttest.js:220]: 👍

File: tests/js/blockchainTest.scrypttest.js [hunk 26]
```diff
@@ -0,0 +1,266 @@
...
+            buildMerklePath(merklePath),
+            buildMerklePath(merklePathOfLastTxCopy),
+            new Sha256(toLittleIndian(Last_TX_ID)),
+            3,
+            preimage).verify()
+
+        expect(result.success, result.error).to.be.false
+    });
+
+    it('testBlockTxCount should FAIL when using NOT last tx', () => {
```
Comment (xhliu) [inline tests/js/blockchainTest.scrypttest.js:242]: NOT the

File: tests/js/blockPRNG.scrypttest.js [hunk 27]
```diff
@@ -0,0 +1,185 @@
...
+let header = {
+    version: 536870912,
+    previousblockhash: "00000000000000000012ce7f4bbce3346ac438ab7fdcb6fa5440db9857856a7f",
+    merkleroot: "95a920b1002bed05379a0d2650bb13eb216138f28ee80172f4cf21048528dc60",
+    time: 1553501874,
+    bits: "180978d5",
+    nonce: 2482491775,
+    difficulty: 116078424449.9863
+}
+
```
Comment (xhliu) [inline tests/js/blockPRNG.scrypttest.js:61]: Similar code in other scrypttest files of this PR. Put in a shared file. Do not duplicate large chunks of code.

File: tests/js/blockPRNG.scrypttest.js [hunk 28]
```diff
@@ -0,0 +1,185 @@
...
+            merkleRoot: new Sha256(toLittleIndian(header.merkleroot)),
+            time: new Bytes(uint322bin(header.time)),
+            bits: new Bytes(toLittleIndian(header.bits)),
+            nonce: new Bytes(uint322bin(header.nonce))
+        }), buildMerklePath(wrongMerklePath), sig, preimage).verify()
+
+        expect(result.success, result.error).to.be.false
+    });
+
+
```
Comment (xhliu) [inline tests/js/blockPRNG.scrypttest.js:184]: Add an e2e test

- when nonce is odd, Bob should NOT win and not able to unlock

**Review by xhliu — CHANGES_REQUESTED** [review]
Very good overall 💯 . Thanks for testing all these, lots of work.

See comments inline.

High-level comments:

- Whenever big piece of code is duplicated, put them in a shared file/function
- Test as many cases as possible.

File: contracts/merklePath.scrypt [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+    Sha256 hash;
+    // invalid {0} or left {1} or right {2} sibling
+    int left;
+}
+
+type MerklePaths = Sibling[MerklePath.DEPTH];
+
+library MerklePath {
+    // N is length of a Sibling
+    static const int N = 33;
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply contracts/merklePath.scrypt:12]: removed

File: tests/js/blockPRNG.scrypttest.js (hunk 16 above)
Comment (zhfnjust) [reply tests/js/blockPRNG.scrypttest.js:6]: move to blockchainhelper, helper.js have a `newTx`, which use random txid

File: contracts/blockchainTest.scrypt [hunk 30]
```diff
@@ -3,36 +3,17 @@ import "blockchain.scrypt";
...
-
-        // verify previous tx is in the block
-        require(Blockchain.txInBlock(prevTxid, bh, merklePath));
-
-        require(Blockchain.blockTxCount(bh, lastTxid, merklePathOfLastTx)  == count);
+    public function testIsBlockHeaderValid(BlockHeader bh,  int blockchainTarget) {
+        require(Blockchain.isBlockHeaderValid(bh, blockchainTarget));
     }
+
+
```
Comment (xhliu) [inline contracts/blockchainTest.scrypt:18]: Add a case to test a chain of headers: 1) they are actually chained one after another; 2) NOT chained

File: tests/js/blockTimeBet.scrypttest.js [hunk 31]
```diff
@@ -194,75 +40,44 @@ describe('Test sCrypt contract blockTimeBet In Javascript', () => {
 
         preimage = getPreimage(tx, blockTimeBet.lockingScript, inputSatoshis)
 
-        const sig = signTx(tx, alicePrivateKey, blockTimeBet.lockingScript, inputSatoshis);
+        const sig = signTx(tx, privateKey, blockTimeBet.lockingScript, inputSatoshis);
 
-        const result = blockTimeBet.main(headers.map(h => toBlockHeader(h)), buildMerklePath(merklePath), sig, preimage).verify()
+        return blockTimeBet.main(headers,
+            proof, sig, preimage).verify()
```
Comment (xhliu) [inline tests/js/blockTimeBet.scrypttest.js:46]: Same line. Format

**Review by xhliu — APPROVED** [review]
💯 Looks great.

Add 1 last test before merging.

**Pull request merged**